File: src/main/java/com/example/demo/service/JwtService.java
package com.example.demo.service;

import io.jsonwebtoken.*;
import io.jsonwebtoken.security.Keys;
import org.springframework.stereotype.Service;

import java.util.Date;

import javax.crypto.SecretKey;

@Service
public class JwtService {

    private static final String SECRET = "supersecretkeysupersecretkeysupersecretkey1234";
    private static final long EXPIRATION = 1000 * 60 * 60; // 1 time

    private final SecretKey key = Keys.hmacShaKeyFor(SECRET.getBytes());

    public String generateToken(String username) {
        return Jwts.builder()
                .subject(username)
                .issuedAt(new Date())
                .expiration(new Date(System.currentTimeMillis() + EXPIRATION))
                .signWith(key)
                .compact();
    }

    public String extractUsername(String token) {
        return parse(token).getSubject();
    }

    public boolean validate(String token) {
        try {
            parse(token);
            return true;
        } catch (JwtException e) {
            return false;
        }
    }

    private Claims parse(String token) {
        return Jwts.parser()
            .verifyWith(key)
            .build()
            .parseSignedClaims(token)
            .getPayload(); 
        }
}

File: src/main/java/com/example/demo/service/JwtService.kt
package com.example.demo.service

import io.jsonwebtoken.Claims
import io.jsonwebtoken.JwtException
import io.jsonwebtoken.Jwts
import io.jsonwebtoken.security.Keys
import org.springframework.stereotype.Service
import java.util.*
import javax.crypto.SecretKey

@Service
open class JwtService {
    private val key: SecretKey = Keys.hmacShaKeyFor(SECRET.toByteArray())

    fun generateToken(username: String?): String? {
        return Jwts.builder()
            .subject(username)
            .issuedAt(Date())
            .expiration(Date(System.currentTimeMillis() + EXPIRATION))
            .signWith(key)
            .compact()
    }

    fun extractUsername(token: String?): String {
        return parse(token)?.subject ?: throw RuntimeException("token not found")
    }

    fun validate(token: String?): Boolean {
        try {
            parse(token)
            return true
        } catch (e: JwtException) {
            return false
        }
    }

    private fun parse(token: String?): Claims? {
        return Jwts.parser()
            .verifyWith(key)
            .build()
            .parseSignedClaims(token)
            .getPayload()
    }

    companion object {
        private const val SECRET = "supersecretkeysupersecretkeysupersecretkey1234"
        private val EXPIRATION = (1000 * 60 * 60 // 1 time
                ).toLong()
    }
}
